File: frontend/src/types/interview.ts
export interface JobPosition {
  id: string
  workspace_id: string
  knowledge_base_id: string | null
  title: string
  department: string | null
  description: string | null
  requirements: Record<string, unknown>
  status: 'DRAFT' | 'ACTIVE' | 'CLOSED'
  created_by: string
  created_at: string
  updated_at: string
}

export interface JobPositionCreateRequest {
  title: string
  department?: string
  description?: string
  knowledge_base_id?: string
  status: 'DRAFT' | 'ACTIVE'
}

export interface JobPositionUpdateRequest {
  title?: string
  department?: string
  description?: string
  knowledge_base_id?: string | null
  status?: 'DRAFT' | 'ACTIVE' | 'CLOSED'
}

export interface CandidateProfile {
  id: string
  workspace_id: string
  user_id: string | null
  resume_knowledge_base_id: string | null
  resume_document_id: string | null
  full_name: string
  email: string | null
  phone: string | null
  source: 'PERSONAL_ACCOUNT' | 'ENTERPRISE_IMPORT'
  status: 'ACTIVE' | 'ARCHIVED'
  profile_data: Record<string, unknown>
  created_by: string
  created_at: string
  updated_at: string
}

export interface CandidateProfileCreateRequest {
  full_name: string
  email?: string
  phone?: string
  resume_document_id: string
}

export interface CandidateProfileUpdateRequest {
  full_name?: string
  email?: string
  phone?: string
  resume_document_id?: string | null
  status?: 'ACTIVE' | 'ARCHIVED'
}

export interface InterviewSession {
  id: string
  workspace_id: string
  job_position_id: string
  job_title: string
  candidate_profile_id: string
  candidate_name: string
  interviewer_id: string | null
  application_id: string | null
  mode: 'MOCK' | 'ENTERPRISE'
  status: string
  current_question_order: number
  configuration: Record<string, unknown>
  scheduled_at: string | null
  started_at: string | null
  completed_at: string | null
  created_by: string
  created_at: string
  updated_at: string
}

export interface InterviewSessionCreateRequest {
  job_position_id: string
  candidate_profile_id: string
  application_id?: string
  reference_knowledge_base_ids?: string[]
  max_question_count?: number
  question_time_limit_minutes?: number
}

export interface InterviewQuestion {
  id: string
  order_no: number
  question_type: string
  content: string
  competency: string | null
  difficulty: 'EASY' | 'MEDIUM' | 'HARD'
  generated_by: string
  status: string
  max_score: number
  expected_points: string[]
  source_evidence: Array<Record<string, unknown>>
  decision_metadata: Record<string, unknown>
}

export interface InterviewPlan {
  id: string
  interview_session_id: string
  version: number
  status: 'DRAFT' | 'READY' | 'FAILED'
  objectives: string[]
  sections: Array<Record<string, unknown>>
  model_name: string | null
  prompt_version: string | null
  generated_at: string | null
  error_message: string | null
  created_at: string
  updated_at: string
  questions: InterviewQuestion[]
}

export interface AIStageMetrics {
  sample_count: number
  average_latency_ms: number
  p50_latency_ms: number
  p95_latency_ms: number
  max_latency_ms: number
}

export interface InterviewObservability {
  interview_id: string
  question_count: number
  measured_turn_count: number
  total_tokens: number
  fallback_event_count: number
  fallback_turn_count: number
  fallback_turn_rate: number
  bottleneck_stage: string | null
  stage_metrics: Record<string, AIStageMetrics>
  route_counts: Record<string, number>
}

export interface InterviewRuntimeQuestion {
  id: string
  order_no: number
  question_type: string
  content: string
  competency: string | null
  difficulty: 'EASY' | 'MEDIUM' | 'HARD'
  generated_by: 'PLAN' | 'FOLLOW_UP' | 'HUMAN'
  asked_at: string | null
}

export interface InterviewTurnCritique {
  id: string
  interview_question_id: string
  score: number
  strengths: string[]
  knowledge_gaps: string[]
  answer_evidence: string[]
  next_action: 'FOLLOW_UP' | 'INCREASE_DIFFICULTY' | 'DECREASE_DIFFICULTY' | 'SWITCH_TOPIC' | 'END_INTERVIEW'
  difficulty_delta: -1 | 0 | 1
  confidence: number
  reason: string
  decision_source: 'MODEL' | 'FALLBACK_RULE'
  model_name: string | null
  prompt_version: string
  created_at: string
}

export interface InterviewPlanRevision {
  id: string
  source_critique_id: string
  version: number
  action: InterviewTurnCritique['next_action']
  target_competency: string | null
  target_difficulty: 'EASY' | 'MEDIUM' | 'HARD' | null
  covered_competencies: string[]
  priority_competencies: string[]
  knowledge_gaps: string[]
  rationale: string
  workflow_trace: Array<Record<string, unknown>>
  before_snapshot: Record<string, unknown>
  after_snapshot: Record<string, unknown>
  change_set: Record<string, { before: unknown; after: unknown }>
  remaining_question_budget: number
  competency_budget: Record<string, number>
  created_at: string
}

export interface InterviewRuntime {
  interview_id: string
  status: string
  current_question: InterviewRuntimeQuestion | null
  completed_question_count: number
  total_question_count: number
  max_question_count: number
  question_time_limit_seconds: number | null
  started_at: string | null
  completed_at: string | null
  follow_up_generated: boolean
  question_timed_out: boolean
  last_turn_feedback: InterviewTurnCritique | null
  adaptive_plan_version: number | null
  evaluation_status: 'PENDING' | 'GENERATING' | 'COMPLETED' | 'FAILED' | null
  decision: 'HIRED' | 'REJECTED' | null
  decided_at: string | null
}

export interface InterviewAnswerSubmitRequest {
  content: string
  duration_seconds?: number
  client_metadata?: Record<string, unknown>
}

export interface InterviewEvaluation {
  id: string
  interview_session_id: string
  status: 'PENDING' | 'GENERATING' | 'COMPLETED' | 'FAILED'
  overall_score: number | null
  dimension_scores: Record<string, number>
  strengths: string[]
  weaknesses: string[]
  evidence: Array<{
    evidence_id: number
    question_id: string
    question: string
    answer_excerpt: string
    dimension: string
    score: number
    finding: string
  }>
  report_text: string | null
  recommendation: string | null
  model_name: string | null
  prompt_version: string | null
  error_message: string | null
  turn_critiques: InterviewTurnCritique[]
  plan_revisions: InterviewPlanRevision[]
  reviewed_at: string | null
  created_at: string
  updated_at: string
}

export interface InterviewQualityGate {
  key: string
  label: string
  value: number | boolean
  threshold: number | boolean
  passed: boolean
  required: boolean
}

export interface InterviewQualityAudit {
  id: string
  interview_session_id: string
  audit_version: string
  passed: boolean
  metrics: Record<string, number | boolean | null>
  quality_gates: InterviewQualityGate[]
  warnings: string[]
  generated_at: string
  created_at: string
}

File: backend/app/schemas/interview.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


JobStatus = Literal["DRAFT", "ACTIVE", "CLOSED"]
CandidateStatus = Literal["ACTIVE", "ARCHIVED"]


class JobPositionCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    department: str | None = Field(default=None, max_length=150)
    description: str | None = None
    requirements: dict = Field(default_factory=dict)
    knowledge_base_id: uuid.UUID | None = None
    status: JobStatus = "DRAFT"


class JobPositionUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    department: str | None = Field(default=None, max_length=150)
    description: str | None = None
    requirements: dict | None = None
    knowledge_base_id: uuid.UUID | None = None
    status: JobStatus | None = None


class JobPositionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    knowledge_base_id: uuid.UUID | None
    title: str
    department: str | None
    description: str | None
    requirements: dict
    status: str
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime


class CandidateProfileCreateRequest(BaseModel):
    full_name: str = Field(min_length=1, max_length=150)
    email: str | None = Field(default=None, max_length=255)
    phone: str | None = Field(default=None, max_length=50)
    resume_document_id: uuid.UUID
    profile_data: dict = Field(default_factory=dict)


class CandidateProfileUpdateRequest(BaseModel):
    full_name: str | None = Field(default=None, min_length=1, max_length=150)
    email: str | None = Field(default=None, max_length=255)
    phone: str | None = Field(default=None, max_length=50)
    resume_document_id: uuid.UUID | None = None
    profile_data: dict | None = None
    status: CandidateStatus | None = None


class CandidateProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    user_id: uuid.UUID | None
    resume_knowledge_base_id: uuid.UUID | None
    resume_document_id: uuid.UUID | None
    full_name: str
    email: str | None
    phone: str | None
    source: str
    status: str
    profile_data: dict
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime


class InterviewSessionCreateRequest(BaseModel):
    job_position_id: uuid.UUID
    candidate_profile_id: uuid.UUID
    application_id: uuid.UUID | None = None
    interviewer_id: uuid.UUID | None = None
    scheduled_at: datetime | None = None
    configuration: dict = Field(default_factory=dict)
    reference_knowledge_base_ids: list[uuid.UUID] = Field(default_factory=list, max_length=5)
    max_question_count: int = Field(default=10, ge=3, le=20)
    question_time_limit_minutes: int = Field(default=10, ge=0, le=60)


class InterviewSessionUpdateRequest(BaseModel):
    interviewer_id: uuid.UUID | None = None
    scheduled_at: datetime | None = None
    configuration: dict | None = None
    status: Literal["DRAFT", "CANCELLED"] | None = None


class InterviewSessionResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    job_position_id: uuid.UUID
    job_title: str
    candidate_profile_id: uuid.UUID
    candidate_name: str
    interviewer_id: uuid.UUID | None
    application_id: uuid.UUID | None
    mode: str
    status: str
    current_question_order: int
    configuration: dict
    scheduled_at: datetime | None
    started_at: datetime | None
    completed_at: datetime | None
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime


class InterviewQuestionResponse(BaseModel):
    id: uuid.UUID
    order_no: int
    question_type: str
    content: str
    competency: str | None
    difficulty: str
    generated_by: str
    status: str
    max_score: float
    expected_points: list
    source_evidence: list
    decision_metadata: dict


class InterviewPlanResponse(BaseModel):
    id: uuid.UUID
    interview_session_id: uuid.UUID
    version: int
    status: str
    objectives: list
    sections: list
    model_name: str | None
    prompt_version: str | None
    generated_at: datetime | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    questions: list[InterviewQuestionResponse]


class AIStageMetricsResponse(BaseModel):
    sample_count: int
    average_latency_ms: int
    p50_latency_ms: int
    p95_latency_ms: int
    max_latency_ms: int


class InterviewObservabilityResponse(BaseModel):
    interview_id: uuid.UUID
    question_count: int
    measured_turn_count: int
    total_tokens: int
    fallback_event_count: int
    fallback_turn_count: int
    fallback_turn_rate: float
    bottleneck_stage: str | None
    stage_metrics: dict[str, AIStageMetricsResponse]
    route_counts: dict[str, int]


class InterviewRuntimeQuestionResponse(BaseModel):
    id: uuid.UUID
    order_no: int
    question_type: str
    content: str
    competency: str | None
    difficulty: str
    generated_by: str
    asked_at: datetime | None


class InterviewTurnCritiqueResponse(BaseModel):
    id: uuid.UUID
    interview_question_id: uuid.UUID
    score: float
    strengths: list[str]
    knowledge_gaps: list[str]
    answer_evidence: list[str]
    next_action: str
    difficulty_delta: int
    confidence: float
    reason: str
    decision_source: str
    model_name: str | None
    prompt_version: str
    created_at: datetime


class InterviewPlanRevisionResponse(BaseModel):
    id: uuid.UUID
    source_critique_id: uuid.UUID
    version: int
    action: str
    target_competency: str | None
    target_difficulty: str | None
    covered_competencies: list[str]
    priority_competencies: list[str]
    knowledge_gaps: list[str]
    rationale: str
    workflow_trace: list[dict]
    before_snapshot: dict
    after_snapshot: dict
    change_set: dict
    remaining_question_budget: int
    competency_budget: dict[str, int]
    created_at: datetime


class InterviewRuntimeResponse(BaseModel):
    interview_id: uuid.UUID
    status: str
    current_question: InterviewRuntimeQuestionResponse | None
    completed_question_count: int
    total_question_count: int
    max_question_count: int
    question_time_limit_seconds: int | None
    started_at: datetime | None
    completed_at: datetime | None
    follow_up_generated: bool = False
    question_timed_out: bool = False
    last_turn_feedback: InterviewTurnCritiqueResponse | None = None
    adaptive_plan_version: int | None = None
    evaluation_status: str | None = None
    decision: str | None = None
    decided_at: datetime | None = None


class InterviewAnswerSubmitRequest(BaseModel):
    content: str = Field(min_length=1, max_length=20000)
    duration_seconds: int | None = Field(default=None, ge=0)
    client_metadata: dict = Field(default_factory=dict)


class InterviewEvaluationResponse(BaseModel):
    id: uuid.UUID
    interview_session_id: uuid.UUID
    status: str
    overall_score: float | None
    dimension_scores: dict
    strengths: list
    weaknesses: list
    evidence: list
    report_text: str | None
    recommendation: str | None
    model_name: str | None
    prompt_version: str | None
    error_message: str | None
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    turn_critiques: list[InterviewTurnCritiqueResponse] = Field(default_factory=list)
    plan_revisions: list[InterviewPlanRevisionResponse] = Field(default_factory=list)


class InterviewQualityAuditResponse(BaseModel):
    id: uuid.UUID
    interview_session_id: uuid.UUID
    audit_version: str
    passed: bool
    metrics: dict
    quality_gates: list[dict]
    warnings: list[str]
    generated_at: datetime
    created_at: datetime
